Classifier and some semantic model for ref var

jcouv opened the pull request:
**Customer scenario**

Use the `ref` locals feature in VS. The color on the `var` keyword is wrong, and hovering over `var` doesn't reveal the inferred type.

**Bugs this fixes:** 
[VSO 168348](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=294604&triage=true)

**Workarounds, if any**
The feature remains usable, but the user cannot see the inferred type.

**Risk**
**Performance impact**
Low. This is a minor change to the classifier and semantic model.

**Is this a regression from a previous update?**
No.

**Root cause analysis:**
Seems like a test gap.

**How was the bug found?**
Customer reporting and ad-hoc testing

jcouv commented:
@dotnet/roslyn-compiler for review.

CyrusNajmabadi commented:
IDE side looks good.

VSadov commented:
LGTM

jcouv commented:
@dotnet/roslyn-compiler for a second code review, please. Small change ;-)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs:
```diff
@@ -1600,5 +1598,33 @@ static void M()
+    static void M()
+    {
+        int i = 0;
+        ref var x = ref i;
```
>ref var x = ref i; [](start = 8, length = 18)

Do we have similar test for ```ref int x = ref i;```?

AlekseyTs replied:
Consider also testing GetAliasInfo API.

---
In reply to: [95439359](https://github.com/dotnet/roslyn/pull/16342#discussion_r95439359) [](ancestors = 95439359)

jcouv replied:
Not that I could find.
Last commit push in this PR now adds one.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs:
```diff
@@ -1600,5 +1598,64 @@ static void M()
+            var type = refVar.Type;
+            Assert.Equal("System.Int32", model.GetTypeInfo(type).Type.ToTestDisplayString());
+            Assert.Equal("System.Int32", model.GetSymbolInfo(type).Symbol.ToTestDisplayString());
+            Assert.Null(model.GetAliasInfo(type));
```
>Assert.Null(model.GetAliasInfo(type)); [](start = 12, length = 38)

It would be good to also have a test when expectation is not null.

jcouv commented:
@AlekseyTs To follow up on why `var` behaved differently than `System.Int32` case.

The node map for `ref type x = i;` has:
- the whole thing
- `ref type`
- `i`

In the case were type is `System.Int32` the node was not found, but unlike `var` we'd be able to bind it on it's own. 
My latest change avoids that and relies on what we find in the node map. If you ask about `type`, we'll use `ref type` to look up in the node map. If the node map is empty, we'll bind the statement and end up with those 3 nodes.

jcouv commented:
@VSadov @AlekseyTs I reworked the PR from scratch, based on our discussion.

With this change `ref var x = i;` binds to 3 nodes:
- the whole thing
- `var` (without the `ref`)
- `i`

The semantic model allows you to ask type and symbol for `var` (or any type in that position).  
It does not support asking about `ref var`.

In passing, I ran into a broken assertion with adding diagnostics when binding an expression that is `SyntaxKind.RefType` (in `BindExpressionInternal`). The problem was that you can't use a token in a diagnostic, you need to use an `IMessageSerializable`, a string or a primitive type (see `AssertMessageSerializable`). I fixed it, but I don't know how to make it visible in tests.

jcouv commented:
@jaredpar Based on bug bar for RTW, I'm going to move this to post-dev15 instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -634,14 +634,14 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
                     {
                         var firstToken = node.GetFirstToken();
-                        diagnostics.Add(ErrorCode.ERR_UnexpectedToken, firstToken.GetLocation(), firstToken);
+                        diagnostics.Add(ErrorCode.ERR_UnexpectedToken, firstToken.GetLocation(), firstToken.ToString());
```
>.ToString() [](start = 107, length = 11)

Not sure what ToString does with trivia, perhaps ValueText instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -598,8 +598,7 @@ private TypeSymbol BindVariableType(CSharpSyntaxNode declarationNode, Diagnostic
             // we want to treat the declaration as an explicitly typed declaration.
 
-            RefKind refKind;
-            TypeSymbol declType = BindType(typeSyntax.SkipRef(out refKind), diagnostics, out isVar, out alias);
```
>typeSyntax.SkipRef(out refKind) [](start = 43, length = 31)

Are we certain it is safe to remove this code? There are multiple callers and it doesn't look like all of them are skipping the ref.

jcouv replied:
None of the other paths are skipping the ref, but I don't think they can have a ref.
The other paths are: binding deconstruction variables, declaration expressions, out vars, out variable declarations. 
The only one that can have ref type is the binding of declaration statement, from what I understand.

AlekseyTs commented:
LGTM

jcouv commented:
@dotnet-bot test windows_eta_open_prtest please
// 12:12:49 
Shims\DTE\CommandsShim.cs(5,7): error CS0246: The type or namespace name 'EnvDTE' could not be found (are you missing a using directive or an assembly reference?)

VSadov commented:
LGTM

cston commented:
@dotnet-bot test windows_eta_open_prtest please

jcouv commented:
Thanks @cston for fixing the ETA/merge problem :-)
